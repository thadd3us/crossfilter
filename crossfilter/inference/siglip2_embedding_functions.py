"""
SigLIP2 embedding functions for images and text.

References:
- SigLIP2 Blog Post: https://huggingface.co/blog/siglip2
- SigLIP2 Model: https://huggingface.co/google/siglip-so400m-patch14-384
"""

import logging
from pathlib import Path

import numpy as np
import torch
from PIL import Image
from transformers import AutoModel, AutoProcessor

logger = logging.getLogger(__name__)


def _get_device() -> torch.device:
    """Get the appropriate device for computation (MPS for Apple Silicon, CUDA for NVIDIA, CPU fallback)."""
    if torch.backends.mps.is_available():
        return torch.device("mps")
    elif torch.cuda.is_available():
        return torch.device("cuda")
    else:
        return torch.device("cpu")


def _center_crop_to_square(image: Image.Image) -> Image.Image:
    """
    Center crop the image to the largest square that fits within the original dimensions.
    This ensures no distortion when resizing to the model's expected input size.
    """
    width, height = image.size
    min_dimension = min(width, height)

    # Calculate center crop coordinates
    left = (width - min_dimension) // 2
    top = (height - min_dimension) // 2
    right = left + min_dimension
    bottom = top + min_dimension

    return image.crop((left, top, right, bottom))


def compute_image_embeddings(
    image_paths: list[Path],
    batch_size: int = 8,
    model_name: str = "google/siglip-so400m-patch14-384",
) -> list[np.ndarray]:
    """
    Compute SigLIP2 embeddings for a list of image paths.

    The function handles images of any size or aspect ratio by finding the largest
    square region that can be extracted without distortion, then resizing that
    region to the model's expected input size.

    Args:
        image_paths: List of paths to image files
        batch_size: Number of images to process in each batch
        model_name: HuggingFace model identifier for SigLIP2

    Returns:
        List of 1D numpy arrays, one embedding vector per input image

    Raises:
        FileNotFoundError: If any image path doesn't exist
        ValueError: If any image cannot be loaded
    """
    # Handle empty input
    if not image_paths:
        logger.info("No images provided, returning empty list")
        return []

    device = _get_device()
    logger.info(f"Using device: {device}")

    # Load model and processor
    model = AutoModel.from_pretrained(model_name, local_files_only=True)
    # TODO: Use torchvision: SiglipImageProcessorFast requires the Torchvision library but it was not found in your environment.
    processor = AutoProcessor.from_pretrained(
        model_name, local_files_only=True, use_fast=False
    )
    model = model.to(device)
    model.eval()

    embeddings = []

    # Process images in batches
    # THAD: TODO: remove the looping logic here -- make this function just for a single batch.
    for i in range(0, len(image_paths), batch_size):
        batch_paths = image_paths[i : i + batch_size]
        batch_images = []

        # Load and preprocess images in the batch
        for image_path in batch_paths:
            if not image_path.exists():
                raise FileNotFoundError(f"Image not found: {image_path}")

            try:
                # Load image and convert to RGB if necessary
                image = Image.open(image_path).convert("RGB")

                # Center crop to square to avoid distortion
                image = _center_crop_to_square(image)

                batch_images.append(image)

            except Exception as e:
                raise ValueError(f"Failed to load image {image_path}: {e}")

        # Process batch through model
        with torch.no_grad():
            inputs = processor(images=batch_images, return_tensors="pt").to(device)

            # Get image embeddings
            outputs = model.get_image_features(**inputs)

            # Normalize embeddings (SigLIP2 typically uses L2 normalization)
            embeddings_normalized = torch.nn.functional.normalize(outputs, p=2, dim=1)

            # Convert to numpy and add to results
            batch_embeddings = embeddings_normalized.cpu().numpy()
            for j in range(len(batch_images)):
                embeddings.append(batch_embeddings[j])

    logger.info(f"Computed embeddings for {len(embeddings)} images")
    return embeddings


def compute_text_embeddings(
    captions: list[str],
    batch_size: int = 32,
    model_name: str = "google/siglip-so400m-patch14-384",
) -> list[np.ndarray]:
    """
    Compute SigLIP2 embeddings for a list of text captions.

    Args:
        captions: List of text strings to encode
        batch_size: Number of captions to process in each batch
        model_name: HuggingFace model identifier for SigLIP2

    Returns:
        List of 1D numpy arrays, one embedding vector per input caption
    """
    # Handle empty input
    if not captions:
        logger.info("No captions provided, returning empty list")
        return []

    device = _get_device()
    logger.info(f"Using device: {device}")

    # Load model and processor
    model = AutoModel.from_pretrained(model_name, local_files_only=True)
    processor = AutoProcessor.from_pretrained(model_name, local_files_only=True)
    model = model.to(device)
    model.eval()

    embeddings = []

    # Process captions in batches
    # TODO: THAD: Remove the looping logic here -- make this function just be for one batch.
    for i in range(0, len(captions), batch_size):
        batch_captions = captions[i : i + batch_size]

        # Process batch through model
        with torch.no_grad():
            inputs = processor(
                text=batch_captions, return_tensors="pt", padding=True, truncation=True
            ).to(device)

            # Get text embeddings
            outputs = model.get_text_features(**inputs)

            # Normalize embeddings (SigLIP2 typically uses L2 normalization)
            embeddings_normalized = torch.nn.functional.normalize(outputs, p=2, dim=1)

            # Convert to numpy and add to results
            batch_embeddings = embeddings_normalized.cpu().numpy()
            for j in range(len(batch_captions)):
                embeddings.append(batch_embeddings[j])

    logger.info(f"Computed embeddings for {len(embeddings)} captions")
    return embeddings


def generate_captions_from_image_embeddings(
    image_embeddings: list[np.ndarray], batch_size: int = 8, max_length: int = 50
) -> list[str]:
    """
    Generate captions from image embeddings.

    Note: SigLIP2 is primarily designed for contrastive learning rather than
    generative captioning. This function provides a placeholder implementation
    that analyzes embedding characteristics to generate descriptive captions.

    For production use, consider:
    1. Using the original images with a dedicated image-to-text model (e.g., BLIP2)
    2. Training a decoder specifically for SigLIP2 embeddings
    3. Using a multimodal model that supports both embedding and generation

    Args:
        image_embeddings: List of image embedding vectors from SigLIP2
        batch_size: Number of embeddings to process in each batch (for consistency)
        max_length: Maximum length of generated captions (for future use)

    Returns:
        List of generated caption strings
    """
    # Handle empty input
    if not image_embeddings:
        logger.info("No embeddings provided, returning empty list")
        return []

    logger.info(f"Generating captions for {len(image_embeddings)} embeddings")

    captions = []

    for i in range(0, len(image_embeddings), batch_size):
        batch_embeddings = image_embeddings[i : i + batch_size]

        # Analyze embedding characteristics to generate descriptive captions
        for embedding in batch_embeddings:
            # Calculate various embedding statistics
            norm = np.linalg.norm(embedding)
            mean_val = np.mean(embedding)
            std_val = np.std(embedding)
            max_val = np.max(embedding)
            min_val = np.min(embedding)

            # Generate captions based on embedding characteristics
            if norm > 0.95 and std_val > 0.08:
                captions.append(
                    "A complex image with rich visual details and varied features"
                )
            elif norm > 0.9 and mean_val > 0.01:
                captions.append("A detailed image with prominent visual elements")
            elif std_val > 0.1:
                captions.append(
                    "An image with high contrast and diverse visual content"
                )
            elif abs(mean_val) < 0.005 and std_val < 0.05:
                captions.append("A balanced image with subtle visual features")
            elif max_val > 0.2 or min_val < -0.2:
                captions.append("An image with distinctive visual characteristics")
            elif norm > 0.8:
                captions.append("An image with moderate visual complexity")
            else:
                captions.append("A simple image with basic visual elements")

    logger.warning(
        "Caption generation from embeddings uses heuristic analysis. "
        "For production use, consider using original images with a "
        "dedicated image-to-text model like BLIP2 or training a "
        "specific decoder for SigLIP2 embeddings."
    )

    logger.info(f"Generated {len(captions)} captions")
    return captions
