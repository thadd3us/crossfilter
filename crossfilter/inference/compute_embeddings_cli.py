"""A Typer CLI to incrementally compute embeddings for a directory of images and write them to a SQLite DB.

Example command line:
```sh
python -m crossfilter.inference.compute_embeddings_cli \
    --embedding_type SIGLIP2 \
    --input_dir /path/to/images_by_uuid \
    --output_embeddings_db /path/to/output.db \
    --batch_size 10 \
    --recompute_existing_embeddings=false \
    --reproject_umap_embeddings=true

```


How this works:
* Images are expected to be named as "<uuid>.jpg" in any subdirectory of the input directory.
* There's a generic "EMBEDDINGS" table in the output DB for all embedding types.
  * The table's primary key column is "UUID".
  * There's a second column called "EMBEDDING" with a msgpack_numpy encoded 1D numpy array.
* When the CLI starts, it figures out which "<uuid>.jpg" files are missing embeddings in the table (unless `--recompute_existing_embeddings` is set to `true`).
* Embedding computation loop:
    * For the missing embeddings, it computes them in batches of `--batch-size` using the `compute_image_embeddings` function from `siglip2_embedding_functions.py` (refactored to allow for incremental output to a sink, including a DB write queue).
    * The embeddings are enqueued to be written to the DB as they are computed, ideally in parallel with the computation.
    * A TQDM progress bar is shown for the embedding computation loop.
    * When that finishes, we join the DB write queue and wait for all the embeddings to be written to the DB.
* UMAP computation:
    * Once all embeddings are computed and written to the DB, if `--reproject_umap_embeddings` is set, the CLI will reproject all the available embeddings to the UMAP space.
    * The UMAP model is stored into the DB as BLOB in a table called UMAP_MODEL, with a single column called MODEL.
    * The embeddings are stored in a DB table called {embedding_type}_UMAP_HAVERSINE, with columns UUID (primary key), SEMANTIC_EMBEDDING_UMAP_LATITUDE, SEMANTIC_EMBEDDING_UMAP_LONGITUDE.
    * Nice-to-have: we seed the UMAP embedding with existing embeddings, so that they don't move "too much" when new UUIDs are added.
"""

import logging
import pickle
import threading
from pathlib import Path
from queue import Queue
from uuid import UUID

import msgpack
import msgpack_numpy
import pandas as pd
import typer
from sqlalchemy import (
    Column,
    LargeBinary,
    MetaData,
    Table,
    create_engine,
)
from sqlalchemy.engine import Engine
from tqdm import tqdm

from crossfilter.core.schema import (
    EmbeddingType,
    EmbeddingsTables,
    SchemaColumns,
)
from crossfilter.data_ingestion.sqlite_utils import upsert_dataframe_to_sqlite
from crossfilter.inference.run_umap import run_umap_projection

logger = logging.getLogger(__name__)


def _get_embedder_instance(embedding_type: EmbeddingType):
    """Get the appropriate embedder instance based on embedding type."""
    if embedding_type == EmbeddingType.SIGLIP2:
        from crossfilter.inference.siglip2_embedding_functions import (
            SigLIP2Embedder,
        )

        return SigLIP2Embedder()
    elif embedding_type == EmbeddingType.FAKE_EMBEDDING_FOR_TESTING:
        from crossfilter.inference.fake_embedding_functions import (
            FakeEmbedder,
        )

        return FakeEmbedder()
    else:
        raise ValueError(f"Unsupported embedding type: {embedding_type}")


app = typer.Typer(
    help="Compute embeddings for images and store them in SQLite database"
)

# Enable msgpack_numpy for serialization/deserialization
msgpack_numpy.patch()


def _create_database_schema(
    engine: Engine, embedding_type: EmbeddingType
) -> Table:
    """Create database schema and return UMAP model table object."""
    metadata = MetaData()

    # Create UMAP model table
    umap_model_table = Table(
        EmbeddingsTables.UMAP_MODEL,
        metadata,
        Column("MODEL", LargeBinary, nullable=False),
    )

    # Note: Both embeddings table and UMAP haversine table are created automatically by upsert_dataframe_to_sqlite

    # Create all tables
    metadata.create_all(engine)

    return umap_model_table


def _scan_image_files(input_dir: Path) -> dict[str, Path]:
    """Scan input directory for UUID.jpg files and return mapping of UUID -> Path."""
    logger.info(f"Scanning for <uuid>.jpg files in {input_dir}")

    if not input_dir.exists():
        raise FileNotFoundError(f"Input directory not found: {input_dir}")

    uuid_to_path: dict[str, Path] = {}

    # Find all .jpg files recursively
    for jpg_path in input_dir.rglob("*.jpg"):
        # Extract UUID from filename (remove .jpg extension)
        uuid_str = jpg_path.stem
        if uuid_str in uuid_to_path:
            logger.warning(f"Duplicate UUID found in {input_dir}: {uuid_str=}")
            continue
        uuid_to_path[uuid_str] = jpg_path

    logger.info(f"Found {len(uuid_to_path)=} .jpg files")
    return uuid_to_path


def _get_existing_embeddings(engine: Engine) -> set[str]:
    """Get set of UUIDs that already have embeddings in the database."""
    try:
        query = f"SELECT {SchemaColumns.UUID_STRING} FROM {EmbeddingsTables.IMAGE_EMBEDDINGS}"
        df = pd.read_sql(query, engine)
        return set(df[SchemaColumns.UUID_STRING].values)
    except Exception:
        # Table doesn't exist yet, return empty set
        return set()


def _write_embeddings_worker(
    write_queue: Queue,
    output_embeddings_db: Path,
    stop_event: threading.Event,
) -> None:
    """Worker thread that writes embeddings to database using upsert_dataframe_to_sqlite."""

    batch_items = []
    
    while not stop_event.is_set() or not write_queue.empty():
        # Collect batch of items from queue
        while len(batch_items) < 100:  # Batch size
            try:
                item = write_queue.get(timeout=0.1)
                if item is None:  # Sentinel value to stop
                    break
                batch_items.append(item)
                write_queue.task_done()
            except Exception:
                # Timeout or queue empty - process what we have
                break
        
        # Process batch if we have items
        if batch_items:
            try:
                # Create DataFrame from batch items
                uuids = []
                embeddings = []
                
                for uuid_str, embedding in batch_items:
                    uuids.append(uuid_str)
                    embeddings.append(embedding)
                
                batch_df = pd.DataFrame({
                    SchemaColumns.UUID_STRING: uuids,
                    SchemaColumns.SEMANTIC_EMBEDDING: embeddings,
                })
                
                # Use the utility function for upsert
                upsert_dataframe_to_sqlite(
                    batch_df, 
                    output_embeddings_db, 
                    EmbeddingsTables.IMAGE_EMBEDDINGS
                )
                
                logger.debug(f"Wrote batch of {len(batch_items)} embeddings to database")
                batch_items.clear()
                
            except Exception as e:
                logger.error(f"Failed to write batch to database: {e}")
                # Re-raise the exception to avoid silent failures
                # The main thread will handle this appropriately
                raise
        
        # Check for sentinel or stop condition
        if not write_queue.empty():
            continue
        elif stop_event.is_set():
            break


def _compute_embeddings_batch(
    batch_paths: list[Path],
    batch_uuids: list[str],
    write_queue: Queue,
    embedder,
) -> None:
    """Compute embeddings for a batch of images and enqueue them for writing."""
    try:
        # Compute embeddings for the batch using the embedder instance
        embeddings = embedder.compute_image_embeddings(batch_paths)

        # Enqueue each embedding for writing
        for uuid_str, embedding in zip(batch_uuids, embeddings):
            write_queue.put((uuid_str, embedding))

    except Exception as e:
        logger.error(f"Error computing embeddings for batch: {e}")
        raise


def _load_embeddings_from_db(engine: Engine) -> pd.DataFrame:
    """Load all embeddings from database into a DataFrame."""
    
    try:
        query = f"""
        SELECT {SchemaColumns.UUID_STRING}, {SchemaColumns.SEMANTIC_EMBEDDING}
        FROM {EmbeddingsTables.IMAGE_EMBEDDINGS}
        """
        df = pd.read_sql(query, engine)
        
        if df.empty:
            return pd.DataFrame(
                columns=[SchemaColumns.UUID_STRING, SchemaColumns.SEMANTIC_EMBEDDING]
            )
        
        # Deserialize embeddings using map operation
        df[SchemaColumns.SEMANTIC_EMBEDDING] = df[SchemaColumns.SEMANTIC_EMBEDDING].map(
            lambda blob: msgpack.unpackb(blob)
        )
        
        return df
    except Exception:
        # Table doesn't exist yet, return empty DataFrame
        return pd.DataFrame(
            columns=[SchemaColumns.UUID_STRING, SchemaColumns.SEMANTIC_EMBEDDING]
        )


def _store_umap_model(engine, umap_model_table: Table, umap_model) -> None:
    """Store UMAP model in database."""

    # Serialize model with pickle
    model_blob = pickle.dumps(umap_model)

    with engine.connect() as conn:
        # Clear existing model first
        conn.execute(umap_model_table.delete())

        # Insert new model
        conn.execute(umap_model_table.insert().values(MODEL=model_blob))
        conn.commit()


def _upsert_umap_embeddings(output_embeddings_db: Path, embedding_type: EmbeddingType, df: pd.DataFrame) -> None:
    """Upsert UMAP embeddings into the haversine table."""
    
    # Filter to rows that have valid UMAP coordinates
    valid_umap_mask = (
        df[SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LATITUDE].notna() &
        df[SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LONGITUDE].notna()
    )
    
    if not valid_umap_mask.any():
        logger.warning("No valid UMAP coordinates to upsert")
        return
    
    # Select only the columns we need for the UMAP haversine table
    umap_df = df[valid_umap_mask][[
        SchemaColumns.UUID_STRING,
        SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LATITUDE,
        SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LONGITUDE,
    ]].copy()
    
    # Use the existing utility function for upsert
    table_name = f"{embedding_type}_UMAP_HAVERSINE"
    upsert_dataframe_to_sqlite(umap_df, output_embeddings_db, table_name)
    
    logger.info(f"Upserted {len(umap_df)} UMAP embeddings into {table_name} table")


@app.command()
def main(
    embedding_type: EmbeddingType = typer.Option(
        EmbeddingType.SIGLIP2,
        "--embedding_type",
        help="Type of embedding to compute (SIGLIP2 or FAKE_EMBEDDING_FOR_TESTING)",
    ),
    input_dir: Path = typer.Option(
        ..., "--input_dir", help="Directory containing images named as <uuid>.jpg"
    ),
    output_embeddings_db: Path = typer.Option(
        ...,
        "--output_embeddings_db",
        help="Path to SQLite database for storing embeddings",
    ),
    batch_size: int = typer.Option(
        10, "--batch_size", help="Batch size for embedding computation"
    ),
    recompute_existing_embeddings: bool = typer.Option(
        False,
        "--recompute_existing_embeddings/--no-recompute_existing_embeddings",
        help="Whether to recompute embeddings that already exist in the database",
    ),
    reproject_umap_embeddings: bool = typer.Option(
        True,
        "--reproject_umap_embeddings/--no-reproject_umap_embeddings",
        help="Whether to reproject embeddings to UMAP space after computation",
    ),
) -> None:
    """Compute embeddings for images and store them in SQLite database."""

    # Set up logging
    logging.basicConfig(level=logging.INFO)

    logger.info("Starting embedding computation with parameters:")
    logger.info(f"  embedding_type: {embedding_type}")
    logger.info(f"  input_dir: {input_dir}")
    logger.info(f"  output_embeddings_db: {output_embeddings_db}")
    logger.info(f"  batch_size: {batch_size}")
    logger.info(f"  recompute_existing_embeddings: {recompute_existing_embeddings}")
    logger.info(f"  reproject_umap_embeddings: {reproject_umap_embeddings}")

    # Validate input directory
    if not input_dir.exists():
        raise typer.BadParameter(f"Input directory does not exist: {input_dir}")

    # Create output directory if needed
    output_embeddings_db.parent.mkdir(parents=True, exist_ok=True)

    # Create database engine
    engine = create_engine(f"sqlite:///{output_embeddings_db}")

    # Create database schema
    umap_model_table = _create_database_schema(engine, embedding_type)

    # Scan for image files
    uuid_to_path = _scan_image_files(input_dir)

    if not uuid_to_path:
        logger.warning("No valid UUID.jpg files found in input directory")
        return

    # Determine which embeddings need to be computed
    if recompute_existing_embeddings:
        missing_uuids = set(uuid_to_path.keys())
        logger.info(f"Recomputing all {len(missing_uuids)} embeddings")
    else:
        existing_uuids = _get_existing_embeddings(engine)
        missing_uuids = set(uuid_to_path.keys()) - existing_uuids
        logger.info(f"Found {len(existing_uuids)} existing embeddings")
        logger.info(f"Need to compute {len(missing_uuids)} new embeddings")

    if not missing_uuids:
        logger.info("No embeddings need to be computed")
    else:
        # Create embedder instance once (efficient model loading)
        logger.info(f"Initializing embedder for {embedding_type}")
        embedder = _get_embedder_instance(embedding_type)

        # Prepare batches for computation
        missing_uuids_list = list(missing_uuids)
        batches = []

        for i in range(0, len(missing_uuids_list), batch_size):
            batch_uuids = missing_uuids_list[i : i + batch_size]
            batch_paths = [uuid_to_path[uuid_str] for uuid_str in batch_uuids]
            batches.append((batch_paths, batch_uuids))

        # Set up database write queue and worker
        write_queue = Queue()
        stop_event = threading.Event()

        # Start database writer thread
        writer_thread = threading.Thread(
            target=_write_embeddings_worker,
            args=(write_queue, output_embeddings_db, stop_event),
        )
        writer_thread.start()

        try:
            # Compute embeddings in batches with progress bar
            logger.info(f"Computing embeddings in {len(batches)} batches")

            for batch_paths, batch_uuids in tqdm(batches, desc="Computing embeddings"):
                _compute_embeddings_batch(
                    batch_paths, batch_uuids, write_queue, embedder
                )

            # Wait for all writes to complete
            logger.info("Waiting for all embeddings to be written to database...")
            write_queue.join()

        finally:
            # Wait for all items to be processed first
            write_queue.join()
            # Then signal to stop and wait for thread to finish processing remaining items
            write_queue.put(None)  # Sentinel value
            stop_event.set()
            writer_thread.join()

    # Run UMAP projection if requested
    if reproject_umap_embeddings:
        logger.info("Loading embeddings for UMAP projection...")
        df = _load_embeddings_from_db(engine)

        if len(df) < 20:
            logger.warning("Two few embeddings found for UMAP projection.")
        else:
            logger.info(f"Running UMAP projection on {len(df)} embeddings...")

            # Run UMAP projection
            umap_model = run_umap_projection(
                df,
                embedding_column=SchemaColumns.SEMANTIC_EMBEDDING,
                output_lat_column=SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LATITUDE,
                output_lon_column=SchemaColumns.SEMANTIC_EMBEDDING_UMAP_LONGITUDE,
            )
            
            # Upsert the UMAP embeddings for the images into the haversine table
            _upsert_umap_embeddings(output_embeddings_db, embedding_type, df)

            # Store UMAP model in database
            _store_umap_model(engine, umap_model_table, umap_model)
            logger.info("UMAP model stored in database")

    logger.info("Embedding computation completed successfully")


if __name__ == "__main__":
    app()
