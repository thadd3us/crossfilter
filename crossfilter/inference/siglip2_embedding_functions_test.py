"""
Tests for SigLIP2 embedding functions.
"""

import logging
from pathlib import Path
import pandas as pd

import numpy as np
import pytest
from syrupy import SnapshotAssertion
from crossfilter.core.schema import SchemaColumns as C

import scipy.spatial.distance
import plotly.express as px

from crossfilter.inference.siglip2_embedding_functions import (
    SigLIP2Embedder,
)
from tests.util.syrupy_html_snapshot import HTMLSnapshotExtension
from crossfilter.inference.test_fixtures import test_df

# Mark all tests in this file as resource intensive to avoid downloading the SIGLIP2 model in regular test runs
# Use fake_embedding_functions_test.py for fast testing instead
# To run these tests: pytest -m "resource_intensive"
pytestmark = pytest.mark.resource_intensive

logger = logging.getLogger(__name__)


@pytest.fixture(scope="session")
def embedder() -> SigLIP2Embedder:
    """Create SigLIP2Embedder instance for testing."""
    return SigLIP2Embedder()


def test_compute_image_and_text_embeddings_match(
    test_df: pd.DataFrame, embedder: SigLIP2Embedder, snapshot: SnapshotAssertion
) -> None:
    df = test_df
    embedder.compute_image_embeddings(df, C.SOURCE_FILE, "image_embedding")
    # df[C.CAPTION] = ("A photo of " + df[C.CAPTION]).str.lower()
    embedder.compute_text_embeddings(df, C.CAPTION, "text_embedding")

    distances = scipy.spatial.distance.cdist(
        np.stack(df["image_embedding"].values),
        np.stack(df["text_embedding"].values),
        metric="cosine",
    )
    assert distances.shape == (len(df), len(df))
    distances_df = pd.DataFrame(distances, index=df["filename"], columns=df[C.CAPTION])
    fig = px.imshow(distances_df)
    html_content = fig.to_html(
        div_id="test-plot-div",
        include_plotlyjs="cdn",
    )
    assert html_content == snapshot(extension_class=HTMLSnapshotExtension)

    df["image_embedding"] = df["image_embedding"].map(lambda x: np.round(x, 4))
    df["text_embedding"] = df["text_embedding"].map(lambda x: np.round(x, 4))
    assert (
        df[[C.CAPTION, "image_embedding", "text_embedding"]].to_dict(orient="records")
        == snapshot
    )


def test_error_handling_invalid_image(
    test_df: pd.DataFrame,
    embedder: SigLIP2Embedder,
    tmp_path: Path,
    snapshot: SnapshotAssertion,
) -> None:
    """Test error handling for invalid image files."""
    # Create a non-image file
    invalid_file = tmp_path / "not_an_image.jpg"
    invalid_file.write_text("This is not an image")

    test_df = test_df.head(3).copy()
    test_df.loc[1, C.SOURCE_FILE] = invalid_file

    embedder.compute_image_embeddings(test_df, "image_path", "embedding")
    assert test_df[[C.CAPTION, "image_embedding"]].to_dict(orient="records") == snapshot


def test_empty_inputs(embedder: SigLIP2Embedder) -> None:
    """Test handling of empty input DataFrames."""
    # Empty image DataFrame
    empty_df = pd.DataFrame({"image_path": []})
    with pytest.raises(ValueError):
        embedder.compute_image_embeddings(empty_df, "image_path", "embedding")
    assert empty_df.empty  # DataFrame should still be empty

    # Empty text DataFrame
    empty_df = pd.DataFrame({"caption": []})
    with pytest.raises(ValueError):
        embedder.compute_text_embeddings(empty_df, "caption", "embedding")
    assert empty_df.empty  # DataFrame should still be empty


# def test_generate_captions_from_image_embeddings(
#     test_df: pd.DataFrame, snapshot: SnapshotAssertion
# ) -> None:
#     """Test caption generation from image embeddings."""
#     df = test_df
#     image_embeddings = compute_image_embeddings(
#         df[C.SOURCE_FILE].to_list(), batch_size=8
#     )
#     df["generated_caption"] = generate_captions_from_image_embeddings(
#         image_embeddings, batch_size=4
#     )
#     assert df[["filename", "generated_caption"]].to_dict(orient="records") == snapshot
