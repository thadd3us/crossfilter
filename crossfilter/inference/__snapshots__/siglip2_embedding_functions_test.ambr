# serializer version: 1
# name: test_compute_image_embeddings_single_batch
  dict({
    'embedding_dimension': 1152,
    'embedding_dtype': 'float32',
    'embedding_norms': list([
      0.9999998807907104,
      1.0,
      1.0,
      0.9999999403953552,
      1.0,
      0.9999999403953552,
      0.9999999403953552,
      1.0,
      0.9999999403953552,
    ]),
    'embedding_samples': list([
      list([
        0.0009717295179143548,
        -0.02490975894033909,
        -0.004316024016588926,
        0.016267668455839157,
        -0.009822698310017586,
      ]),
      list([
        -0.014484883286058903,
        -0.01488995086401701,
        0.02086106315255165,
        0.01467109564691782,
        -0.003560800803825259,
      ]),
      list([
        -0.021870028227567673,
        -0.02694854885339737,
        0.01908637396991253,
        0.001822814578190446,
        -0.0069392044097185135,
      ]),
      list([
        0.006100810598582029,
        -0.029960324987769127,
        0.0011419188231229782,
        0.007112536579370499,
        -0.006558549124747515,
      ]),
      list([
        0.021284017711877823,
        -0.005231924820691347,
        0.009300594218075275,
        -0.024449441581964493,
        -0.012065443210303783,
      ]),
      list([
        -0.012142828665673733,
        -0.0068155680783092976,
        -0.01857057400047779,
        0.010030635632574558,
        -0.0031028941739350557,
      ]),
      list([
        -0.014416868798434734,
        -0.007140066009014845,
        0.023280374705791473,
        -0.008634159341454506,
        0.004601079039275646,
      ]),
      list([
        0.037984028458595276,
        0.019249849021434784,
        -0.008281292393803596,
        -0.01394366379827261,
        -0.023509258404374123,
      ]),
      list([
        -0.0188488457351923,
        0.02247016876935959,
        0.0245710089802742,
        0.014699585735797882,
        -0.008729813620448112,
      ]),
    ]),
    'embedding_shapes': list([
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
      tuple(
        1152,
      ),
    ]),
    'image_filenames': list([
      '01_golden_gate_bridge.jpg',
      '02_earth_from_space.jpg',
      '03_forest_path.jpg',
      '04_mountain_view.jpg',
      '05_city_skyline.jpg',
      '06_nature_close_up.jpg',
      '07_street_scene.jpg',
      '08_abstract_pattern.jpg',
      '09_urban_details.jpg',
    ]),
    'num_embeddings': 9,
  })
# ---
# name: test_compute_image_embeddings_small_batches
  dict({
    'batch_size_used': 3,
    'embedding_dimension': 1152,
    'embedding_samples': list([
      list([
        0.0009717264329083264,
        -0.024909762665629387,
        -0.004316024947911501,
      ]),
      list([
        -0.014484884217381477,
        -0.014889958314597607,
        0.020861055701971054,
      ]),
      list([
        -0.021870024502277374,
        -0.02694854885339737,
        0.01908637396991253,
      ]),
      list([
        0.006100814789533615,
        -0.029960326850414276,
        0.0011419174261391163,
      ]),
      list([
        0.021284015849232674,
        -0.005231920164078474,
        0.009300600737333298,
      ]),
      list([
        -0.012142826803028584,
        -0.0068155680783092976,
        -0.01857057213783264,
      ]),
      list([
        -0.014416875317692757,
        -0.007140063680708408,
        0.02328038215637207,
      ]),
      list([
        0.037984028458595276,
        0.01924986019730568,
        -0.008281292393803596,
      ]),
      list([
        -0.018848851323127747,
        0.022470176219940186,
        0.0245710089802742,
      ]),
    ]),
    'num_embeddings': 9,
  })
# ---
# name: test_compute_text_embeddings
  dict({
    'embedding_dimension': 1152,
    'embedding_dtype': 'float32',
    'embedding_norms': list([
      0.9999999403953552,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      0.9999999403953552,
      1.0,
    ]),
    'embedding_samples': list([
      list([
        -0.016999298706650734,
        0.002107082400470972,
        0.00017189675418194383,
        -0.01337739173322916,
        0.01832893118262291,
      ]),
      list([
        -0.011260532774031162,
        -0.0244089774787426,
        -0.009199096821248531,
        -0.03165854886174202,
        -0.01092254277318716,
      ]),
      list([
        0.009701629169285297,
        -0.017985614016652107,
        0.003903348930180073,
        -0.009664807468652725,
        -0.0064050969667732716,
      ]),
      list([
        0.0005202649044804275,
        0.017131593078374863,
        0.0471951849758625,
        -0.006504232529550791,
        -0.005245743785053492,
      ]),
      list([
        -0.030400894582271576,
        -0.009422089904546738,
        0.004443463869392872,
        -0.03535837680101395,
        -0.008314130827784538,
      ]),
      list([
        -0.02228523977100849,
        0.017967550083994865,
        0.026143966242671013,
        0.02379930205643177,
        0.006802549120038748,
      ]),
      list([
        -0.015376145951449871,
        -0.0038125391583889723,
        -0.01878773793578148,
        -0.03732398897409439,
        -0.03511198237538338,
      ]),
      list([
        0.007444205693900585,
        -0.013728169724345207,
        0.03517366573214531,
        -0.04165700450539589,
        -0.025379473343491554,
      ]),
      list([
        -0.016109798103570938,
        0.015486950054764748,
        0.016431298106908798,
        -0.03525971621274948,
        -0.015682023018598557,
      ]),
    ]),
    'input_captions': list([
      'A beautiful golden gate bridge in San Francisco',
      'Planet Earth as seen from space showing blue oceans and white clouds',
      'A peaceful forest path with tall trees on both sides',
      'Stunning mountain landscape with snow-capped peaks',
      'Modern city skyline with tall skyscrapers at sunset',
      'Close-up view of colorful flowers in nature',
      'Busy street scene with people and vehicles',
      'Abstract geometric pattern with vibrant colors',
      'Urban architectural details and building textures',
    ]),
    'num_embeddings': 9,
  })
# ---
# name: test_generate_captions_from_image_embeddings
  dict({
    'caption_lengths': list([
      44,
      44,
      44,
      44,
      44,
      44,
      44,
      44,
      44,
    ]),
    'captions': list([
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
      'A balanced image with subtle visual features',
    ]),
    'image_filenames': list([
      '01_golden_gate_bridge.jpg',
      '02_earth_from_space.jpg',
      '03_forest_path.jpg',
      '04_mountain_view.jpg',
      '05_city_skyline.jpg',
      '06_nature_close_up.jpg',
      '07_street_scene.jpg',
      '08_abstract_pattern.jpg',
      '09_urban_details.jpg',
    ]),
    'num_captions': 9,
  })
# ---
